test/unit/org/apache/cassandra/io/filesystem/ListenableFileSystem.java

```diff
@@ -0,0 +1,541 @@
+            onOpen.add((OnOpen) listener);
+        if (listener instanceof OnRead)
+            onRead.add((OnRead) listener);
+        if (listener instanceof OnTransferTo)
+            onTransferTo.add((OnTransferTo) listener);
+        if (listener instanceof OnWrite)
+            onWrite.add((OnWrite) listener);
+        if (listener instanceof OnTransferFrom)
+            onTransferFrom.add((OnTransferFrom) listener);
+        if (listener instanceof OnChannelMeta)
+            onChannelMeta.add((OnChannelMeta) listener);
+        return () -> remove(listener);
+    }
+
+    public void remove(Listener listener)
+    {
+        lists.forEach(l -> l.remove(listener));
+    }
+
+    public void clearListeners()
+    {
+        lists.forEach(List::clear);
+    }
+
+    private interface ListenerAction<T>
+    {
+        void accept(T value) throws IOException;
+    }
+
+    private <T> void notifyListeners(List<T> listeners, ListenerAction<T> fn) throws IOException
+    {
+        for (T listener : listeners)
+            fn.accept(listener);
+    }
+
+    @Override
+    protected Path wrap(Path p)
+    {
+        return p instanceof ListenablePath ? p : new ListenablePath(p);
+    }
+
+    @Override
+    protected Path unwrap(Path p)
+    {
+        return p instanceof ListenablePath ? ((ListenablePath) p).delegate : p;
+    }
+
+    @Override
+    public ListenableFileSystemProvider provider()
+    {
+        return provider;
+    }
+
+
+    private class ListenableFileSystemProvider extends ForwardingFileSystemProvider
+    {
+        ListenableFileSystemProvider(FileSystemProvider delegate)
+        {
+            super(delegate);
+        }
+
+        @Override
+        protected Path wrap(Path a)
+        {
+            return ListenableFileSystem.this.wrap(a);
+        }
+
+        @Override
+        protected Path unwrap(Path p)
+        {
+            return ListenableFileSystem.this.unwrap(p);
+        }
+
+        @Override
+        public OutputStream newOutputStream(Path path, OpenOption... options) throws IOException
+        {
+            int len = options.length;
+            Set<OpenOption> opts = new HashSet<>(len + 3);
+            if (len == 0)
+            {
+                opts.add(StandardOpenOption.CREATE);
+                opts.add(StandardOpenOption.TRUNCATE_EXISTING);
+            }
+            else
+            {
+                for (OpenOption opt : options)
+                {
+                    if (opt == StandardOpenOption.READ)
+                        throw new IllegalArgumentException("READ not allowed");
+                    opts.add(opt);
+                }
+            }
+            opts.add(StandardOpenOption.WRITE);
+            return Channels.newOutputStream(newFileChannel(path, opts));
+        }
+
+        @Override
+        public InputStream newInputStream(Path path, OpenOption... options) throws IOException
+        {
+            if (options.length > 0)
```

Comment:
nit: if check not strictly necessary

Reply:
was copy/paste from java... prob to avoid iterators?

Reply:
Depends on how smart the runtime is I guess?